Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp

```diff
@@ -0,0 +1,475 @@
+{
+    if (m_detached_widgets_window.is_null()) {
+        m_detached_widgets_window = TRY(GUI::Window::try_create());
+        m_detached_widgets_window->set_title(section_label().to_deprecated_string());
+        m_detached_widgets_window->set_window_type(WindowType::Normal);
+        if (m_detached_size.has_value())
+            m_detached_widgets_window->resize(detached_size());
+        else
+            m_detached_widgets_window->resize(640, 480);
+
+        m_detached_widgets_window->center_on_screen();
+
+        auto root_container = m_detached_widgets_window->set_main_widget<GUI::Frame>();
+        root_container->set_fill_with_background_color(true);
+        root_container->set_layout<GUI::VerticalBoxLayout>();
+        root_container->set_frame_style(Gfx::FrameStyle::Window);
+        auto transfer_childs = [this](auto reciever, auto childs) {
+            for (RefPtr<GUI::Widget> widget : childs) {
+                if (widget.ptr() == this->m_controls_spacer)
+                    continue;
+                widget->remove_from_parent();
+                widget->set_visible(true);
+                (void)reciever->try_add_child(*widget);
+            }
+        };
+
+        transfer_childs(root_container, child_widgets());
+
+        m_detached_widgets_window->on_close = [this, root_container, transfer_childs]() {
+            transfer_childs(this, root_container->child_widgets());
+            this->set_view_state(ViewState::Expanded);
+            unregister_open_window(m_detached_widgets_window);
+            this->m_detached_widgets_window = nullptr;
+        };
+    }
+
+    register_open_window(m_detached_widgets_window);
+
+    if (m_is_dragging)
+        m_detached_widgets_window->move_to(screen_relative_rect().location().translated(m_current_mouse_position).translated({ m_detached_widgets_window->width() / -2, 0 }));
+    m_detached_widgets_window->show();
+    return {};
+}
+
+void DynamicWidgetContainer::close_all_detached_windows()
+{
+    for (auto window : DynamicWidgetContainer::s_open_windows.in_reverse()) {
+        window->close();
+    }
+}
+
+void DynamicWidgetContainer::register_open_window(RefPtr<GUI::Window> window)
+{
+    DynamicWidgetContainer::s_open_windows.append(window);
+}
+
+void DynamicWidgetContainer::unregister_open_window(RefPtr<GUI::Window> window)
+{
+    Optional<size_t> match = DynamicWidgetContainer::s_open_windows.find_first_index(window);
+    if (match.has_value())
+        DynamicWidgetContainer::s_open_windows.remove(match.value());
+}
+
+void DynamicWidgetContainer::mousedown_event(MouseEvent& event)
+{
+    if (event.button() != MouseButton::Primary)
+        return;
+
+    if (m_expand_collapse_button.contains(event.position()) && !is_container_with_individual_order())
+        set_view_state(view_state() == ViewState::Expanded ? ViewState::Collapsed : ViewState::Expanded);
+
+    if (m_detach_widget_button.contains(event.position()))
+        set_view_state(ViewState::Detached);
```

Comment:
A more general thing that I was asking myself only after a while of looking at this code: Is there a reason why you're reinventing buttons and labeling and various GUI functionality and layouting calculations for the container controls? It would be much better if you could add a widget that contains the label and the control buttons. That would also allow us to change the primary direction of the container to horizontal instead of vertical, since that only requires to change the widget's layout type to vertical instead of horizontal. I think you can also apply the knurling better in this way.

Reply:
There was a reason, initially i was aiming to have the buttons drawn as a overlay over the containing widgets and therefore i started to draw them manually as a "on_second_paint" thing. But that did not work out as expected in regards to the mouse events getting catched or propagted to the childs. With that idea the DynamicWidgetContainer would not consume space for itself.
With the current solution the "header" is taking up some space... i'll refactor that to make use of existing button and label widgets.

Reply:
A new widget that contains the label and buttons has been implemented.